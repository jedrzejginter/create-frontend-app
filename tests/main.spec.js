const fs = require("fs");
const path = require("path");
const glob = require("glob");

const OUT_DIR = process.env.PROJECT_OUT;
const { PROJECT_NAME } = process.env;

function outPath(...segments) {
  return path.join(process.cwd(), OUT_DIR, ...segments);
}

function readOutFile(...segments) {
  return fs.readFileSync(outPath(...segments), "utf-8");
}

// eslint-disable-next-line import/no-dynamic-require
const pkg = require(require.resolve(outPath("package.json")));

test("should have correct files structure", () => {
  const projectPaths = glob
    .sync(outPath("**/*"), {
      nodir: true,
      dot: true,
    })
    .map((p) => path.relative(outPath(), p));

  expect(projectPaths).toContain(".github/workflows/cd.yml");
  expect(projectPaths).toContain(".github/workflows/ci.yml");
  expect(projectPaths).toContain(".vscode/settings.json");
  expect(projectPaths).toContain("babel/plugin-resolve-imports.js");
  expect(projectPaths).toContain("pages/404.tsx");
  expect(projectPaths).toContain("pages/_app.ts");
  expect(projectPaths).toContain("pages/_document.ts");
  expect(projectPaths).toContain("pages/dashboard.tsx");
  expect(projectPaths).toContain("pages/index.ts");
  expect(projectPaths).toContain("pages/login.ts");
  expect(projectPaths).toContain("pages/create-account/activate.ts");
  expect(projectPaths).toContain("pages/create-account/index.ts");
  expect(projectPaths).toContain("pages/forgot-password/index.ts");
  expect(projectPaths).toContain("pages/forgot-password/reset.ts");
  expect(projectPaths).toContain("public/mockServiceWorker.js");
  expect(projectPaths).toContain("public/robots.txt");
  expect(projectPaths).toContain("scripts/lint-package-json.js");
  expect(projectPaths).toContain("scripts/rewrite-package-json.js");
  expect(projectPaths).toContain("scripts/setup.sh");
  expect(projectPaths).toContain("src/assets/icons/home.svg");
  expect(projectPaths).toContain("src/components/FormError/FormError.tsx");
  expect(projectPaths).toContain("src/components/FormError/index.ts");
  expect(projectPaths).toContain("src/components/Spinner/Spinner.tsx");
  expect(projectPaths).toContain("src/components/Spinner/index.ts");
  expect(projectPaths).toContain("src/containers/Session/Session.tsx");
  expect(projectPaths).toContain("src/containers/Session/index.ts");
  expect(projectPaths).toContain("src/next/App/App.tsx");
  expect(projectPaths).toContain("src/next/App/index.ts");
  expect(projectPaths).toContain("src/next/Document/Document.tsx");
  expect(projectPaths).toContain("src/next/Document/index.ts");
  expect(projectPaths).toContain("src/pages/404/404.tsx");
  expect(projectPaths).toContain("src/pages/404/index.ts");
  expect(projectPaths).toContain(
    "src/pages/ActivateAccount/ActivateAccount.tsx"
  );
  expect(projectPaths).toContain("src/pages/ActivateAccount/index.ts");
  expect(projectPaths).toContain("src/pages/CreateAccount/CreateAccount.tsx");
  expect(projectPaths).toContain("src/pages/CreateAccount/index.ts");
  expect(projectPaths).toContain("src/pages/CreateAccount/types.ts");
  expect(projectPaths).toContain("src/pages/Dashboard/Dashboard.tsx");
  expect(projectPaths).toContain("src/pages/Dashboard/index.ts");
  expect(projectPaths).toContain("src/pages/ForgotPassword/ForgotPassword.tsx");
  expect(projectPaths).toContain("src/pages/ForgotPassword/index.ts");
  expect(projectPaths).toContain("src/pages/ForgotPassword/types.ts");
  expect(projectPaths).toContain("src/pages/Home/Home.tsx");
  expect(projectPaths).toContain("src/pages/Home/index.ts");
  expect(projectPaths).toContain("src/pages/Login/Login.tsx");
  expect(projectPaths).toContain("src/pages/Login/index.ts");
  expect(projectPaths).toContain("src/pages/Login/types.ts");
  expect(projectPaths).toContain("src/pages/ResetPassword/ResetPassword.tsx");
  expect(projectPaths).toContain("src/pages/ResetPassword/index.ts");
  expect(projectPaths).toContain("src/pages/ResetPassword/types.ts");
  expect(projectPaths).toContain("src/services/api/client.ts");
  expect(projectPaths).toContain("src/services/api/index.ts");
  expect(projectPaths).toContain("src/services/api/utils.ts");
  expect(projectPaths).toContain("src/services/auth/api.ts");
  expect(projectPaths).toContain("src/services/auth/const.ts");
  expect(projectPaths).toContain("src/services/auth/index.ts");
  expect(projectPaths).toContain("src/services/auth/utils.ts");
  expect(projectPaths).toContain("src/services/mock-server/browser.ts");
  expect(projectPaths).toContain("src/services/mock-server/handlers.ts");
  expect(projectPaths).toContain("src/services/mock-server/index.ts");
  expect(projectPaths).toContain("src/services/mock-server/server.ts");
  expect(projectPaths).toContain("src/types/core.ts");
  expect(projectPaths).toContain("src/utils/misc.ts");
  expect(projectPaths).toContain("typings/inline-react-svg.d.ts");
  expect(projectPaths).toContain("typings/node.d.ts");
  expect(projectPaths).toContain(".dockerignore");
  expect(projectPaths).toContain(".env.example");
  expect(projectPaths).toContain(".eslintignore");
  expect(projectPaths).toContain(".gitignore");
  expect(projectPaths).toContain(".nvmrc");
  expect(projectPaths).toContain("Dockerfile");
  expect(projectPaths).toContain("README.md");
  expect(projectPaths).toContain("babel.config.js");
  expect(projectPaths).toContain("eslint.config.js");
  expect(projectPaths).toContain("jest.config.js");
  expect(projectPaths).toContain("jest.setup.js");
  expect(projectPaths).toContain("lint-staged.config.js");
  expect(projectPaths).toContain("next-env.d.ts");
  expect(projectPaths).toContain("next.config.js");
  expect(projectPaths).toContain("package.json");
  expect(projectPaths).toContain("prettier.config.js");
  expect(projectPaths).toContain("tsconfig.eslint.json");
  expect(projectPaths).toContain("tsconfig.json");
  expect(projectPaths).toContain("utils.js");
  expect(projectPaths).toContain("yarn.lock");

  expect(projectPaths).not.toContain(".env");
});

describe(".gitignore", () => {
  it("should not ignore typings directory", () => {
    expect(readOutFile(".gitignore")).toEqual(
      expect.not.stringMatching(/^typings/im)
    );
  });
});

describe("package.json", () => {
  it("should have project name set", () => {
    expect(pkg.name).toBe(PROJECT_NAME);
  });

  it("should have husky config set", () => {
    expect(pkg).not.toHaveProperty("__husky");
    expect(pkg).toHaveProperty("husky");
  });

  it("should have all packages with exact versions", () => {
    const inexact = [];
    const deps = { ...pkg.dependencies, ...pkg.devDependencies };

    for (const dependency in deps) {
      if (!/^[\da-z]/.test(deps[dependency])) {
        inexact.push(dependency);
      }
    }

    expect(inexact).toStrictEqual([]);
  });
});
