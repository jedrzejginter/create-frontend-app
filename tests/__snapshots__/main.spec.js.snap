// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have correct files structure 1`] = `
Array [
  ".vscode/settings.json",
  ".vscode/snippets.code-snippets",
  "babel/plugin-resolve-imports.js",
  "pages/404.tsx",
  "pages/_app.ts",
  "pages/_document.ts",
  "pages/dashboard.tsx",
  "pages/index.ts",
  "pages/login.ts",
  "pages/create-account/activate.ts",
  "pages/create-account/index.ts",
  "pages/forgot-password/index.ts",
  "pages/forgot-password/reset.ts",
  "public/mockServiceWorker.js",
  "public/robots.txt",
  "scripts/lint-package-json.js",
  "scripts/rewrite-package-json.js",
  "src/assets/icons/home.svg",
  "src/components/FormError/FormError.tsx",
  "src/components/FormError/index.ts",
  "src/components/Spinner/Spinner.tsx",
  "src/components/Spinner/index.ts",
  "src/containers/Session/Session.tsx",
  "src/containers/Session/index.ts",
  "src/next/App/App.tsx",
  "src/next/App/index.ts",
  "src/next/Document/Document.tsx",
  "src/next/Document/index.ts",
  "src/pages/404/404.tsx",
  "src/pages/404/index.ts",
  "src/pages/ActivateAccount/ActivateAccount.tsx",
  "src/pages/ActivateAccount/index.ts",
  "src/pages/CreateAccount/CreateAccount.tsx",
  "src/pages/CreateAccount/index.ts",
  "src/pages/CreateAccount/types.ts",
  "src/pages/Dashboard/Dashboard.tsx",
  "src/pages/Dashboard/index.ts",
  "src/pages/ForgotPassword/ForgotPassword.tsx",
  "src/pages/ForgotPassword/index.ts",
  "src/pages/ForgotPassword/types.ts",
  "src/pages/Home/Home.tsx",
  "src/pages/Home/index.ts",
  "src/pages/Login/Login.tsx",
  "src/pages/Login/index.ts",
  "src/pages/Login/types.ts",
  "src/pages/ResetPassword/ResetPassword.tsx",
  "src/pages/ResetPassword/index.ts",
  "src/pages/ResetPassword/types.ts",
  "src/services/api/client.ts",
  "src/services/api/index.ts",
  "src/services/api/utils.ts",
  "src/services/auth/api.ts",
  "src/services/auth/const.ts",
  "src/services/auth/index.ts",
  "src/services/auth/utils.ts",
  "src/services/mock-server/browser.ts",
  "src/services/mock-server/handlers.ts",
  "src/services/mock-server/index.ts",
  "src/services/mock-server/server.ts",
  "src/types/core.ts",
  "src/types/react.ts",
  "src/utils/misc.ts",
  "typings/inline-react-svg.d.ts",
  "typings/node.d.ts",
  ".dockerignore",
  ".env.example",
  ".eslintignore",
  ".gitignore",
  ".nvmrc",
  "Dockerfile",
  "README.md",
  "babel.config.js",
  "eslint.config.js",
  "jest.config.js",
  "jest.setup.js",
  "lint-staged.config.js",
  "next-env.d.ts",
  "next.config.js",
  "package.json",
  "prettier.config.js",
  "tsconfig.eslint.json",
  "tsconfig.json",
  "utils.js",
  "yarn.lock",
]
`;
